tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs

```diff
@@ -17,144 +20,232 @@ namespace Datadog.Trace.Agent.DiscoveryService
+            : this(apiRequestFactory, initialRetryDelayMs: 500, maxRetryDelayMs: 5_000, recheckIntervalMs: 30_000)
+        {
+        }
 
-        public bool? ClientDropP0s { get; private set;  }
+        /// <summary>
+        /// Gets all the supported endpoints for testing purposes only
+        /// </summary>
+        public static string[] AllSupportedEndpoints =>
+            new[]
+            {
+                SupportedDebuggerEndpoint,
+                SupportedConfigurationEndpoint,
+                SupportedStatsEndpoint,
+                SupportedDataStreamsEndpoint,
+            };
 
         public static DiscoveryService Create(ImmutableExporterSettings exporterSettings)
-        {
-            lock (GlobalLock)
-            {
-                var apiRequestFactory = AgentTransportStrategy.Get(
+            => new(
+                AgentTransportStrategy.Get(
                     exporterSettings,
                     productName: "discovery",
                     tcpTimeout: TimeSpan.FromSeconds(15),
                     AgentHttpHeaderNames.MinimalHeaders,
                     () => new MinimalAgentHeaderHelper(),
-                    uri => uri);
+                    uri => uri));
+
+        /// <inheritdoc cref="IDiscoveryService.SubscribeToChanges"/>
+        public void SubscribeToChanges(Action<AgentConfiguration> callback)
+        {
+            lock (_lock)
+            {
+                if (!_agentChangeCallbacks.Contains(callback))
+                {
+                    _agentChangeCallbacks.Add(callback);
+                }
+            }
+
+            if (Volatile.Read(ref _configuration) is { } currentConfig)
+            {
+                // If we already have fetched the config, call this immediately
+                callback(currentConfig);
+            }
+        }
+
+        /// <inheritdoc cref="IDiscoveryService.RemoveSubscription"/>
+        public void RemoveSubscription(Action<AgentConfiguration> callback)
+        {
+            lock (_lock)
+            {
+                _agentChangeCallbacks.Remove(callback);
+            }
+        }
+
+        private void NotifySubscribers(AgentConfiguration newConfig)
+        {
+            List<Action<AgentConfiguration>> subscribers;
+            lock (_lock)
+            {
+                subscribers = _agentChangeCallbacks.ToList();
+                // Setting the configuration immediately after grabbing
+                // the subscribers ensures subscribers receive the
+                // notification exactly once
+                _configuration = newConfig;
+            }
 
-                // TODO: If custom instrumentation updates the exporterSettings, we need to refresh the DiscoveryService instance and make sure all users get updated
-                return Instance ??= new DiscoveryService(apiRequestFactory);
+            foreach (var subscriber in subscribers)
+            {
+                subscriber(newConfig);
```

Comment:
I would add `try..catch` block around this.

Reply:
I'm actually in two minds 🤔 We already have a try-catch around the whole lot, so we're "safe" from that point of view. Adding try-catch for every subscriber adds overhead... and these should be very simple callbacks. But then, if one subscriber has an error, it would prevent the others from updating which seems bad... I'm torn

Reply:
Changed my mind to agree with you, added in https://github.com/DataDog/dd-trace-dotnet/commit/860e6809f7004743bbc84a72f6b7b7e84e2ba3bd

Reply:
FWIW I agree with @shurivich here, I think we should prefer an error handling policy wherein one subscriber failing would not cause another subscriber to never receive the notification, since each subscriber is completely independent and their functionality would most likely work just fine regardless of the other subscribers.

I don't think the overhead is really a big issue here, since we control the polling interval, and it's not something we are doing super often.